1) Сделал контроллер UserController в котором добавил

DmitryKarpushov opened the pull request:
-Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,49 +1,94 @@
-    private Integer generateId() {
-        idTask++;
-        return idTask;
+    private final int countPopularFilms = 10;
```
Значение по-умолчанию можно указать в самой аннотации @RequestParam

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,49 +1,94 @@
+    @GetMapping()
+    public List<Film> getFilms() {
+        log.info("FilmController. getFilms. Получаем все фильмы.");
+        return filmService.getFilmsList();
```
`getFilmsList` не указываем тип в названиях -> getFilms()

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,49 +1,94 @@
+    @PutMapping()
+    public ResponseEntity<Film> updateFilm(@Valid @RequestBody @NotNull Film film) {
+        log.info("FilmController. updateFilm. Обновление фильма.");
+        if (!filmService.getFilms().containsKey(film.getId())) {
```
filmService.getFilms().containsKey(film.getId() а это всё должно быть в хранилище как метод который возвращает фильмы/сущности по id
Этого метода у тебя нет! 

А фильмы как Map не должны отдаваться наружу.

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -14,43 +17,77 @@
+    @GetMapping("/{id}")
+    public ResponseEntity<User> getUser(@PathVariable int id) {
+        log.info("UserController. getUser. Получение пользователя.");
+        if (!userService.getUsers().containsKey(id)) {
```
Тут всё тоже самое, просто подсвечу...

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/datavalidation/ValidationFieldsFilm.java:
```diff
@@ -1,24 +1,20 @@
         if (!films.containsKey(film.getId())) {
-            logger.info("ValidationFieldsFilm.isValidDate: Проверяем на существование фильма");
-            throw new NotFoundException("Такого фильма не существует");
+            log.info("ValidationFieldsFilm.isValidDate: Проверяем на существование фильма");
```
давай в лог больше полезной информации...

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/datavalidation/ValidationFieldsUser.java:
```diff
@@ -1,26 +1,22 @@
 
-    private static final Logger logger = LoggerFactory.getLogger(UserController.class);
-
     public static void noFoundUser(User user, Map<Integer, User> users) {
```
Но вообще подобные валидации не нужны... всё должно быть проще, спросил у хранилища сущность по id и если её нет кидаем ексепшн.

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/exception/FilmNotFoundException.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * @author Дмитрий Карпушов 18.12.2022
+ */
+public class FilmNotFoundException extends RuntimeException {
```
Эксепшн должен быть один на все сущности, не нужно создавать для каждой сущности классы. Просто `NotFoundException`

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/handler/ErrorHandler.java:
```diff
@@ -0,0 +1,48 @@
+ * @author Дмитрий Карпушов 05.12.2022
+ */
+@RestControllerAdvice
+public class ErrorHandler {
```
Тут не хватает логирования...

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmStorage.java:
```diff
@@ -0,0 +1,20 @@
+public interface FilmStorage {
+    Map<Integer, Film> getFilms();
+
+    Film addFilm(Film film);
```
Методы не висят в воздухе!
Поэтому нет нужды в них повторять название сущности, например: `filmStorage.addFilm()`

Получается масло-масляное! 

Лучше: `filmStorage.add()` `filmStorage.getAll()` `filmStorage.getById()`

FanaticFiz reviewed (changes requested):
Финальный комментарий на Я.Практикуме

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,49 +1,74 @@
+    @GetMapping()
+    public List<Film> getFilms() {
+        log.info("FilmController. getFilms. Получаем все фильмы.");
+        return new ArrayList<>(filmService.getFilms().values());
```
Как `List` отдаём еще начиная с хранилища

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,49 +1,74 @@
+    @GetMapping("/{id}")
+    public ResponseEntity<Film> getFilm(@PathVariable int id) {
+        log.info("FilmController. getFilm. Получение фильма.");
+        return new ResponseEntity<>(filmService.getFilmById(id), HttpStatus.OK);
```
Не прячем "главные" действия куда то в строку...
Оно достойно отдельной строки кода.
`filmService.getFilmById(id)`

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    private Boolean existsFilm(Integer filmId) {
+        if (!getFilms().containsKey(filmId)) {
```
И это тоже перенести в хранилище... 
Метод isExist() отлично его дополнит. 
Но обычно всё же просто спрашивают сущность по id и делают из этого выводы.

Но совершенно точно не нужно отдавать фильмы как карту наружу!

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,117 @@
+
+    public User addUser(User user) {
+        log.info("UserService.addUser: Начали создание пользователя");
+        if (user.getName() == null || user.getName().isEmpty() || user.getName().isBlank()) {
```
Кажется у тебя теперь дублируется это... оно же есть и в методе обновления и конструкторе в User.

Надо чтобы было написано один раз!

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmStorage.java:
```diff
@@ -0,0 +1,23 @@
+@Component
+public interface FilmStorage {
+
+    Map<Integer, Film> getAll();
```
Map<Integer, Film> getAll(); -> List<Film> getAll();

FanaticFiz reviewed (changes requested):
Финальный комментарий на Я.Практикуме

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
 5)Переработка Архитектуры. Вынесли хранение данных о фильмах и пользователях в отдельные классы.
 6)Обеспечили возможность пользователям добавлять друг друга в друзья и ставить фильмам лайки.

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,82 @@
+    InMemoryFilmStorage inMemoryFilmStorage;
+
+    @Autowired
+    public FilmService(InMemoryFilmStorage inMemoryFilmStorage) {
```
Подключить нужно было бы именно через интерфейс, иначе зачем же они нам нужны! ))

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmStorage.java:
```diff
@@ -0,0 +1,23 @@
+
+    List<Film> getAll();
+
+    Film getById(Integer id);
```
Далее мы будем осваивать Optional - можешь заранее почитать про него...
